internal/auth/auth.go

```diff
@@ -0,0 +1,172 @@
+package auth
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+	"net/url"
+	"os"
+
+	"github.com/pkg/browser"
+	"github.com/speakeasy-api/speakeasy/internal/config"
+)
+
+const (
+	// Replace these with a local server for development
+	appURL = "https://app.speakeasyapi.dev"
+	apiURL = "https://api.prod.speakeasyapi.dev"
+)
+
+type authResult struct {
+	config.SpeakeasyAuthInfo
+	err error
+}
+
+func Authenticate(force bool) error {
+	ctx := context.Background()
+
+	existingApiKey, setByEnvVar := config.GetSpeakeasyAPIKey()
+	if existingApiKey != "" && !force {
+		if err := testAuth(existingApiKey); err != nil {
+			if setByEnvVar {
+				return err
+			}
+		} else {
+			return nil
+		}
+	}
+
+	if !force {
+		fmt.Println("Authentication needed")
+	}
+
+	done := make(chan authResult)
+
+	addr, srv := startServer(done)
+	defer func() { _ = srv.Shutdown(ctx) }()
+
+	hostname, err := os.Hostname()
+	if err != nil {
+		return err
+	}
+
+	url := fmt.Sprintf("%s?cli_callback_url=%s&cli_host_name=%s\n", appURL, addr, hostname)
+
+	if err := browser.OpenURL(url); err != nil {
+		fmt.Println("Please open the following URL in your browser:", url)
+	}
+
+	res := <-done
+	if res.err != nil {
+		return res.err
+	}
+
+	if err := testAuth(res.APIKey); err != nil {
+		return err
+	}
+
+	if err := config.SetSpeakeasyAuthInfo(res.SpeakeasyAuthInfo); err != nil {
+		return fmt.Errorf("failed to save API key: %w", err)
+	}
+
+	fmt.Println("Authentication successful!")
```

Comment:
Could we mention the workspace that the user has successfully authenticated with + a link to the general workspace login page ?

Reply:
I currently don't have access to the workspace name will a link to the workspace be enough?

Reply:
@simplesagar have added the workspace url